### PR TITLE
Add `loop()` table function

### LINKED ISSUE #52765: Table function `loop`
**Use case**

Tests and benchmarks.

```
SELECT ... FROM loop(database, table);
SELECT ... FROM loop(database.table);
SELECT ... FROM loop(table);
SELECT ... FROM loop(other_table_function(...));
```

**Describe the solution you'd like**

Every interface method is forwarded, but for reading, the pipeline is modified to repeat the reading when it is finished,
and writing is not allowed. Trivial count optimization is disabled. Index analysis works in the same way as for the argument's table.

It is unspecified if the function should use the same table snapshot on every reading or different.

### PR DESCRIPTION
**Pull request by sarielwxm**
Fixes #52765  

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added a new table function `loop` to support returning query results in an infinite loop.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>Modify your CI run</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  All with Azure
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

<details>

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 3933c59a289b134b8f5af944e373ec088dbcd0b2 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63452/3933c59a289b134b8f5af944e373ec088dbcd0b2/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by divanik** [comment]
Hi, @sarielwxm. Thank you for your contribution to Clickhouse! I will be glad to review your PR. Firstly, I'd like to ask you to add the example output after execution of this table function that demonstrates its typical behaviour to the documentation. Secondly, could you please [write a fast test](https://clickhouse.com/docs/en/development/tests) that checks desired behaviour of the function? After these steps I will be able to continue the reviewing process

**Comment by sarielwxm** [comment]
> Hi, @sarielwxm. Thank you for your contribution to Clickhouse! I will be glad to review your PR. Firstly, I'd like to ask you to add the example output after execution of this table function that demonstrates its typical behaviour to the documentation. Secondly, could you please [write a fast test](https://clickhouse.com/docs/en/development/tests) that checks desired behaviour of the function? After these steps I will be able to continue the reviewing process

Sure. I have already tested it locally. But I have a problem because this table function implements an infinite loop over the query results and it seems difficult to represent the example output. Could you please give me some suggestions? Much appreciated!

**Comment by divanik** [comment]
> Sure. I have already tested it locally. But I have a problem because this table function implements an infinite loop over the query results and it seems difficult to represent the example output. Could you please give me some suggestions? Much appreciated!

Could you use LIMIT operator to bound the output of the loop function?

**Comment by sarielwxm** [comment]
> Could you use LIMIT operator to bound the output of the loop function?

Got it. It's useful.

File: src/TableFunctions/TableFunctionLoop.cpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        StorageID(getDatabaseName(), table_name),
+        storage
+        );
+    res->startup();
+    return res;
+}
+
+void registerTableFunctionLoop(TableFunctionFactory & factory)
+{
+    factory.registerFunction<TableFunctionLoop>();
```
Comment (divanik) [inline src/TableFunctions/TableFunctionLoop.cpp:141]: Could you please add description here as it is done here: https://github.com/ClickHouse/ClickHouse/blob/03b47ddcbdf84f5af39775f52051352e17fe26f5/src/TableFunctions/TableFunctionAzureBlobStorage.cpp#L388-L389

File: contrib/azure
Comment (divanik) [inline contrib/azure:1]: It seems to me that your changes have nothing in common with Azure. Could you revert this changes from this PR?

File: docs/en/sql-reference/table-functions/loop.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+``` sql
+SELECT * FROM loop(test_database, test_table);
+SELECT * FROM loop(test_database.test_table);
+SELECT * FROM loop(test_table);
+```
+
+Or using other table function:
+
+``` sql
+SELECT * FROM loop(numbers(10));
```
Comment (divanik) [inline docs/en/sql-reference/table-functions/loop.md:37]: Please add work result of this function as it is done in the functional test. For instance, SELECT * FROM loop(numbers(3)) LIMIT 7.

File: src/TableFunctions/TableFunctionLoop.cpp [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    if (args.empty())
+        throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "No arguments provided for table function 'loop'");
+
+    if (args.size() == 1)
+    {
+        if (const auto * id = args[0]->as<ASTIdentifier>())
+        {
+            String id_name = id->name();
+
+            size_t dot_pos = id_name.find('.');
```
Comment (divanik) [inline src/TableFunctions/TableFunctionLoop.cpp:61]: Let's check that there is no more than one dot.

File: src/Processors/QueryPlan/ReadFromLoopStep.cpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                loop = true;
+            }
+            Chunk chunk;
+            if (executor->pull(chunk))
+            {
+                if (chunk)
+                    return chunk;
+            }
+            else
+            {
```
Comment (divanik) [inline src/Processors/QueryPlan/ReadFromLoopStep.cpp:77]: Just a Suggestion: let's add retries here. I mean, if inner_storage failed to pull X times in a row we'd better to fail here not to hang

File: contrib/azure
Comment (divanik) [reply contrib/azure:1]: Also I suspect that the problems with build in CI can be connected with this change.

File: src/Processors/QueryPlan/ReadFromLoopStep.cpp (hunk 4 above)
Comment (sarielwxm) [reply src/Processors/QueryPlan/ReadFromLoopStep.cpp]: Do you mean set a retry count and remake if the pull goes successful?

**Comment by divanik** [comment]
[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63452/056dcd61c4fb07e2893748af4997510191c32ebf/ci_running.html) to open a full report on a separate page - you can find a report, why some CI tests failed. Tests that you added shouldn't be flaky. But there are some flaky tests in our repository (tests that fail in master CI, and they can fail in your PR as well). You can find them at https://aretestsgreenyet.com. If you see that a test fail both in your PR and master, you can forget about it more or less. But in our case, the new test is flaky that's why it should be fixed before PR merge. Everything else looks good to me

**Review by divanik — CHANGES_REQUESTED** [review]
There is need to fix the flaky test

**Review by divanik — APPROVED** [review]
LGTM. There are some issues with CI, I will try to resolve them by myself

**Comment by alexey-milovidov** [comment]
@divanik @sarielwxm The table function `loop` is extremely slow:

`SELECT sum(cityHash64(*)) FROM loop(hits)`

It does processing in a single thread.
The main purpose of the function is testing, and it should be able to work at full speed.

File: src/TableFunctions/TableFunctionLoop.cpp [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+        else
+        {
+            auto inner_table_function = TableFunctionFactory::instance().get(inner_table_function_ast, context);
+            storage = inner_table_function->execute(
+                    inner_table_function_ast,
+                    context,
+                    table_name,
+                    std::move(cached_columns),
+                    is_insert_query);
```
Comment (arthurpassos) [inline src/TableFunctions/TableFunctionLoop.cpp:131]: The 5th argument of `ITableFunction::execute` is `bool use_global_context` and not `is_insert_query`, see https://github.com/ClickHouse/ClickHouse/blob/3b2a325327b63ef25d5546d69f1132fcf1788d7a/src/TableFunctions/ITableFunction.h#L85

Comment (azat) [reply src/TableFunctions/TableFunctionLoop.cpp:131]: Thanks, fixed in https://github.com/ClickHouse/ClickHouse/pull/78707

**Comment by clickgapai** [comment]
Hi @sarielwxm @divanik — while reviewing this PR I found the following:

- **Bug (P2):** `loop()` string-splits identifier on '.', misroutes backtick-quoted dotted table names to a different (or nonexistent) database — https://github.com/ClickHouse/ClickHouse/issues/104175

Happy to discuss — close anything that's wrong or already addressed.
